packages/typescript-checker/src/typescript-checker.ts

```diff
@@ -33,176 +29,149 @@ export function create(injector: Injector<PluginContext>): TypescriptChecker {
+   * These groups will get send to the check method.
+   * @param mutants All the mutants to group.
    */
-  private resolveCheckResult(): void {
-    if (this.currentErrors.length) {
-      const errorText = ts.formatDiagnostics(this.currentErrors, {
-        getCanonicalFileName: (fileName) => fileName,
-        getCurrentDirectory: process.cwd,
-        getNewLine: () => EOL,
-      });
-      this.currentTask.resolve({
-        status: CheckStatus.CompileError,
-        reason: errorText,
-      });
+  public async group(mutants: Mutant[]): Promise<string[][]> {
+    if (this.typeScriptCheckeroptions.typeScriptChecker.strategy === 'noGrouping') {
+      return mutants.map((m) => [m.id]);
     }
-    this.currentTask.resolve({ status: CheckStatus.Passed });
+    const nodes = this.tsCompiler.getFileRelationsAsNodes();
+
+    const mutantsOutSideProject = mutants.filter((m) => nodes.get(toPosixFileName(m.fileName)) == null).map((m) => m.id);
+    const mutantsToTest = mutants.filter((m) => nodes.get(toPosixFileName(m.fileName)) != null);
+
+    const groups = createGroups(mutantsToTest, nodes);
+    const sortedGroups = groups.sort((a, b) => b.length - a.length);
+    const result = mutantsOutSideProject.length ? [mutantsOutSideProject, ...sortedGroups] : sortedGroups;
+
+    this.logger.info(`Created ${result.length} groups with largest group of ${result[0].length} mutants`);
+    return result;
   }
 
-  /**
-   * Clear state between checks
-   */
-  private clearCheckState() {
-    while (this.currentErrors.pop()) {
-      // Idle
+  private async checkErrors(
+    mutants: Mutant[],
+    errorsMap: Record<string, ts.Diagnostic[]>,
+    nodes: Map<string, Node>
+  ): Promise<Record<string, ts.Diagnostic[]>> {
+    const errors = await this.tsCompiler.check(mutants);
+    const mutantsThatCouldNotBeTestedInGroups = new Set<Mutant>();
+
+    //If there is only a single mutant the error has to originate from the single mutant
+    if (errors.length && mutants.length === 1) {
+      errorsMap[mutants[0].id] = errors;
+      return errorsMap;
+    }
+
+    for (const error of errors) {
+      const nodeErrorWasThrownIn = nodes.get(error.file?.fileName ?? '');
+      if (!nodeErrorWasThrownIn) {
+        throw new Error('Error not found in any node');
+      }
+      const mutantsRelatedToError = nodeErrorWasThrownIn.getMutantsWithReferenceToChildrenOrSelf(mutants);
+
+      if (mutantsRelatedToError.length === 1) {
+        if (errorsMap[mutantsRelatedToError[0].id]) {
+          errorsMap[mutantsRelatedToError[0].id].push(error);
+        } else {
+          errorsMap[mutantsRelatedToError[0].id] = [error];
+        }
+      } else if (mutantsRelatedToError.length === 0) {
+        for (const mutant of mutants) {
+          mutantsThatCouldNotBeTestedInGroups.add(mutant);
+        }
+      } else {
+        for (const mutant of mutantsRelatedToError) {
+          mutantsThatCouldNotBeTestedInGroups.add(mutant);
+        }
```

Comment:
Could you add a rule of comment in each of the 3 scenarios here? What does each scenario mean?